api/api-shared-v2/src/main/java/com/thoughtworks/go/apiv2/shared/representers/EnvironmentVariableRepresenter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2019 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv2.shared.representers;
+
+import com.thoughtworks.go.api.base.OutputWriter;
+import com.thoughtworks.go.api.representers.ErrorGetter;
+import com.thoughtworks.go.api.representers.JsonReader;
+import com.thoughtworks.go.api.util.HaltApiResponses;
+import com.thoughtworks.go.apiv2.shared.exception.InvalidGoCipherTextRuntimeException;
+import com.thoughtworks.go.config.EnvironmentVariableConfig;
+import com.thoughtworks.go.config.EnvironmentVariablesConfig;
+import com.thoughtworks.go.security.CryptoException;
+
+import java.util.HashMap;
+import java.util.Optional;
+
+
+public class EnvironmentVariableRepresenter {
+    public static void toJSON(OutputWriter outputWriter, EnvironmentVariableConfig environmentVariableConfig) {
+        if (!environmentVariableConfig.errors().isEmpty()) {
+            outputWriter.addChild("errors", errorWriter -> {
+                HashMap<String, String> mapping = new HashMap<>();
+                mapping.put("encryptedValue", "encrypted_value");
+                new ErrorGetter(mapping).toJSON(outputWriter, environmentVariableConfig);
+            });
+        }
+
+        outputWriter
+                .add("secure", environmentVariableConfig.isSecure())
+                .add("name", environmentVariableConfig.getName());
+        addValue(outputWriter, environmentVariableConfig);
+    }
+
+    public static EnvironmentVariablesConfig fromJSONArray(JsonReader jsonReader) {
+        EnvironmentVariablesConfig variables = new EnvironmentVariablesConfig();
+        jsonReader.readArrayIfPresent("environment_variables", environmentVariables -> {
+            environmentVariables.forEach(variable -> {
+                variables.add(EnvironmentVariableRepresenter.fromJSON(new JsonReader(variable.getAsJsonObject())));
+            });
+        });
+        return variables;
+    }
+
+    public static EnvironmentVariableConfig fromJSON(JsonReader jsonReader) {
+        String name = jsonReader.getString("name");
+        Boolean secure = jsonReader.optBoolean("secure").orElse(false);
+        Optional<String> optValue = jsonReader.optString("value");
+        Optional<String> optEncValue = jsonReader.optString("encrypted_value");
+
+        if (!optValue.isPresent() && !optEncValue.isPresent()) {
```

Comment:
Why perform this check when `environmentVariableConfig.deserialize` handles it? https://github.com/gocd/gocd/blob/master/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentVariableConfig.java#L302

Reply:
`if (!optValue.isPresent() && !optEncValue.isPresent())` is testing if none of the two(encrypted_value, value) are provided, while the test [environmentVariableConfig.deserialize](https://github.com/gocd/gocd/blob/master/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentVariableConfig.java#L302) is testing if both values are provided. These tests are checking the opposite.

Reply:
Its okay to not provide either. There are instances on build.go.cd where just the name exists and no value, so that it can be overridden at the time of scheduling the pipeline - https://build.gocd.org/go/admin/pipelines/PublishStableRelease/environment_variables. I don't remember what happens if `{"name": "foo", "secure": true}` is provided. Maybe that's something to check and handle in `environmentVariableConfig.deserialize`.

Reply:
I think for both pipeline config and environments, environment variable representer should be consistent. And, providing both value=nil and encrypted_value=nil (through API or XML) is not accepted currently (results in https://github.com/gocd/gocd/blob/master/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentVariableConfig.java#L198 when saved via XML, and results in https://github.com/gocd/gocd/blob/master/api/api-base/src/main/java/com/thoughtworks/go/api/util/HaltApiMessages.java#L93 via API, for pipelines). This check is useful. However, if you're doing it here, then the apiv6 EnvironmentVariableRepresenter must reflect the same. Its better to push this check to the EnvironmentVariableConfig#deserialize method then, don't you think?

Reply:
Currently, all APIs accepting environment variables in request body fail with 500 when you don't provide any of the two (`encrypted_value` or `value`) properties, I agree that the fix to this should be consistent across all APIs. I have created an [issue](https://github.com/gocd/gocd/issues/6017) to target this. 

For this PR, I think we following check:

`if (!optValue.isPresent() && !optEncValue.isPresent()) `